torch/distributed/_dist2.py

```diff
@@ -0,0 +1,136 @@
+
+This is an experimental new API for PyTorch Distributed. This is actively in development and subject to change or deletion entirely.
+
+This is intended as a proving ground for more flexible and object oriented distributed APIs.
+"""
+
+from datetime import timedelta
+from typing import Protocol
+
+import torch
+from torch._C._distributed_c10d import Backend, ProcessGroup, Store
+from torch.distributed.rendezvous import rendezvous
+
+
+_BACKENDS: dict[str, "ProcessGroupFactory"] = {}
+
+
+class ProcessGroupFactory(Protocol):
+    """Protocol for process group factories."""
+
+    def __call__(
+        self,
+        store: Store,
+        rank: int,
+        world_size: int,
+        timeout: timedelta,
+        device: torch.device,
+        pg_options: Backend.Options,
+    ) -> ProcessGroup: ...
+
+
+def register_backend(name: str, func: ProcessGroupFactory) -> None:
+    """
+    Register a new process group backend.
+
+    Args:
+        name: The name of the backend.
+        func: The function to create the process group.
+    """
+    if name in _BACKENDS:
+        raise ValueError(f"Backend {name} already registered")
+
+    _BACKENDS[name] = func
+
+
+def _gloo_factory(
+    store: Store,
+    rank: int,
+    world_size: int,
+    timeout: timedelta,
+    device: torch.device,
+    pg_options: Backend.Options,
+) -> ProcessGroup:
+    from torch.distributed import ProcessGroupGloo
+
+    assert pg_options is None, "Gloo backend does not support options"
+
+    backend_class = ProcessGroupGloo(store, rank, world_size, timeout)
+    backend_class._set_sequence_number_for_group()
+
+    pg = ProcessGroup(store, rank, world_size)
+    pg._set_default_backend(ProcessGroup.BackendType.GLOO)
+
+    # register devices
+    pg._register_backend(device, ProcessGroup.BackendType.GLOO, backend_class)
+    pg._register_backend(
+        torch.device("cpu"), ProcessGroup.BackendType.GLOO, backend_class
+    )
+    if torch.cuda.is_available():
+        pg._register_backend(
+            torch.device("cuda"), ProcessGroup.BackendType.GLOO, backend_class
+        )
+    return pg
+
+
+def _nccl_factory(
+    store: Store,
+    rank: int,
+    world_size: int,
+    timeout: timedelta,
+    device: torch.device,
+    pg_options: Backend.Options,
+) -> ProcessGroup:
+    from torch.distributed import ProcessGroupNCCL
+
+    assert isinstance(pg_options, ProcessGroupNCCL.Options)
+
+    pg_options._timeout = timeout
+
+    backend_class = ProcessGroupNCCL(store, rank, world_size, pg_options)
+    backend_class._set_sequence_number_for_group()
+    backend_class.eager_connect_single_device(device)
+
+    pg = ProcessGroup(store, rank, world_size)
+    pg._set_default_backend(ProcessGroup.BackendType.NCCL)
+    pg._register_backend(device, ProcessGroup.BackendType.NCCL, backend_class)
+
+    return pg
+
+
+register_backend("gloo", _gloo_factory)
+register_backend("nccl", _nccl_factory)
+
+
+def new_group(
```

Comment:
previously in init_process_group we always create the default PG first so that function like static globalRank() will return the correct global rank, what is the expected behavior now?

Reply:
With this we're trying to move away from depending on global ranks. Users generally should query PG.rank and PG.world_size instead

Reply:
but we need global ranks for comm split here:

std::shared_ptr<NCCLComm> NCCLComm::split(...)

Internally people need global ranks. I guess for now I can just leave it empty.